### PR TITLE
[Feature] Support pivot syntax

### LINKED ISSUE #20520: function: pivot and unpivot
## Feature request
It's better to have pivot function to support more flexible scenario. By now, we can only use case when to do pivot, and it has bad performance. for unpivot, we can use unnest, but still need an aggregation.
https://learn.microsoft.com/en-us/sql/t-sql/queries/from-using-pivot-and-unpivot?view=sql-server-ver15

### PR DESCRIPTION
**Pull request by macroguo-ghy**
## Why I'm doing:
https://github.com/StarRocks/starrocks/issues/20520 
Support pivot syntax
## What I'm doing:

Fixes https://github.com/StarRocks/starrocks/issues/20520

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 1]
```diff
@@ -313,6 +313,7 @@ PLUGIN: 'PLUGIN';
 PLUGINS: 'PLUGINS';
 PIPE: 'PIPE';
 PIPES: 'PIPES';
+PIVOT: 'PIVOT';
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:316]: please add these key word in `nonReserved` of Starrocks.g4

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java [hunk 2]
```diff
@@ -863,6 +875,116 @@ public Scope visitValues(ValuesRelation node, Scope scope) {
...
         }
 
+        @Override
+        public Scope visitPivotRelation(PivotRelation node, Scope context) {
+            AnalyzeState analyzeState = new AnalyzeState();
+            Scope queryScope = process(node.getQuery(), context);
+
+            List<Pair<String, Expr>> aggAliasExprs = new ArrayList<>();
+            for (PivotAggregation pivotAggregation : node.getAggregateFunctions()) {
+                analyzeExpression(pivotAggregation.getFunctionCallExpr(), analyzeState, queryScope);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java:885]: should we check whether it is a agg function?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/PivotRelation.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        return rewrittenAggFunctions;
+    }
+
+    @Override
+    public <R, C> R accept(AstVisitor<R, C> visitor, C context) {
+        return visitor.visitPivotRelation(this, context);
+    }
+
+    public Map<String, SlotRef> getUsedColumns() {
+        Map<String, SlotRef> map = new HashMap<>();
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/PivotRelation.java:109]: is a case insensitive map more robust?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 4]
```diff
@@ -2077,7 +2077,7 @@ querySpecification
     ;
 
 fromClause
-    : (FROM relations)?                                                                 #from
+    : (FROM relations)? pivotClause?                                                    #from
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2080]: Is pivotClause can exist without a relation?

File: fe/fe-core/src/test/java/com/starrocks/analysis/SelectStmtTest.java [hunk 5]
```diff
@@ -108,6 +109,47 @@ public static void setUp() throws Exception {
...
+                "c0,k3,k4,{'a','a'}_s,{'a','a'}_a,{'b','b'}_s,{'b','b'}_a,{'c','c'}_s,{'c','c'}_a", columns);
+
+        sql = "select * from t0 join tbl1 "
+                + "pivot (sum(t0.c1) as s, avg(t0.c2) as a "
+                + "for (k1, k2) "
+                + "in (('a', 'a') as aa, ('b', 'b') as bb, ('c', 'c') as cc, ('d', 'd') as dd)) order by t0.c0";
+        columns = String.join(",",
+                UtFrameUtils.getPlanAndFragment(starRocksAssert.getCtx(), sql).second.getColNames());
+        Assertions.assertEquals(
+                "c0,k3,k4,aa_s,aa_a,bb_s,bb_a,cc_s,cc_a,dd_s,dd_a", columns);
```
Comment (packy92) [inline fe/fe-core/src/test/java/com/starrocks/analysis/SelectStmtTest.java:150]: Could you add some test to verify the case when result?

Comment (macroguo-ghy) [reply fe/fe-core/src/test/java/com/starrocks/analysis/SelectStmtTest.java:150]: OK, I add them to end2end test.

**Review by macroguo-ghy — COMMENTED** [review]
@packy92 thanks for your review, I have addressed your review, please take a look again.

**Comment by macroguo-ghy** [comment]
@packy92 Can you help me review again?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 4 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2080]: So maybe `(from relations pivotClause?) ?`

**Review by packy92 — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Review by Seaven — COMMENTED** [review]
Fix the .g4 comment plz

**Comment by macroguo-ghy** [comment]
> Fix the .g4 comment plz

Done

**Review by Seaven — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**